### PR TITLE
Add cmdlinet::option_names

### PR DESCRIPTION
**Pull request by hannes-steffenhagen-diffblue**
This function can be used to iterate over the command line arguments in
a for-each loop.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: src/util/cmdline.cpp [hunk 1]
```diff
@@ -247,3 +247,77 @@ bool cmdlinet::parse(int argc, const char **argv, const char *optstring)
...
+  return *this;
+}
+
+const cmdlinet::option_namest::option_names_iteratort
+cmdlinet::option_namest::option_names_iteratort::operator--(int)
+{
+  return --option_names_iteratort(*this);
+}
+
+const std::basic_string<char> cmdlinet::option_namest::option_names_iteratort::
```
Comment (tautschnig) [inline src/util/cmdline.cpp]: Should be `std::string` as used in the declaration of `optstring`

**Review by tautschnig — APPROVED** [review]

**Review by smowton — APPROVED** [review]

File: src/util/cmdline.h [hunk 2]
```diff
@@ -40,6 +41,46 @@ class cmdlinet
     return getoptnr(option).has_value();
   }
 
+  struct option_namest
+  {
+    option_namest(const cmdlinet &command_line);
+    struct option_names_iteratort
+      : public std::iterator<std::bidirectional_iterator_tag, std::string>
+    {
+      explicit option_names_iteratort() = default;
```
Comment (sonodtt) [inline src/util/cmdline.h]: is explicit here not just ignored?
https://stackoverflow.com/questions/6775336/c-explicit-keyword-for-function-without-arguments
(not that stackoverflow is the ultimate reference :), it isn't - but for context here)

Comment (hannes-steffenhagen-diffblue) [reply src/util/cmdline.h]: I think it is, I'll remove it.

File: src/util/cmdline.cpp (hunk 1 above)
Comment (hannes-steffenhagen-diffblue) [reply src/util/cmdline.cpp]: :+1: Thanks, fixing!

**Review by danpoe — APPROVED** [review]
LGTM

File: src/util/cmdline.h [hunk 3]
```diff
@@ -40,6 +41,46 @@ class cmdlinet
     return getoptnr(option).has_value();
   }
 
+  struct option_namest
```
Comment (peterschrammel) [inline src/util/cmdline.h:44]: Why not `class`?

**Review by peterschrammel — APPROVED** [review]

File: src/util/cmdline.h (hunk 3 above)
Comment (hannes-steffenhagen-diffblue) [reply src/util/cmdline.h:44]: No particular reason.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: a915e269dd8fc607008357fcb85e9d15ca4d71cd).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98202762

**Comment by hannes-steffenhagen-diffblue** [comment]
Do not merge for now, I need to think about how to handle options that only have a option character rather than a full option name

**Comment by hannes-steffenhagen-diffblue** [comment]
Fixed/Changed: The iterator now skips over unset options and short options.

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: b1405940a59d96cc5200c8e0a140d93f18355bed).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98336913
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 3e0e19b2975b756a2a411ed279ff11def4ae23a4).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/98339491

**Pull request merged**